src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java

```diff
@@ -73,26 +84,52 @@ public RedirectView login(final @RequestParam(name = REDIRECT_URL_NAME) Optional
                 && customJwtService.isTokenValid(customJWTCookie.getValue());
 
         if (!isTokenValid) {
-            final Cookie redirectUrlCookie = WebUtil.buildSecureCookie(REDIRECT_URL_NAME,
-                    redirectUrlParam.orElse(configProperties.getDefaultRedirectUrl()));
-            response.addCookie(redirectUrlCookie);
+            final String state = encryptionService.getSHA512SecurePassword(
+                    oneLoginService.generateAndStoreState(response, redirectUrlParam.orElse(configProperties.getDefaultRedirectUrl()))
+            );
+            final String nonce = oneLoginService.generateAndStoreNonce();
 
-            return new RedirectView(oneLoginService.getOneLoginAuthorizeUrl());
+            return new RedirectView(oneLoginService.getOneLoginAuthorizeUrl(state, nonce));
         }
 
         return new RedirectView(redirectUrlParam.orElse(configProperties.getDefaultRedirectUrl()));
     }
 
     @GetMapping("/redirect-after-login")
-    public RedirectView redirectAfterLogin(final @CookieValue(name = REDIRECT_URL_NAME) String redirectUrlCookie,
+    public RedirectView redirectAfterLogin(
+            final @CookieValue(name = STATE_COOKIE) String stateCookie,
             final HttpServletResponse response,
-            final @RequestParam String code) {
-        final OneLoginUserInfoDto userInfo = oneLoginService.getOneLoginUserInfoDto(code);
-        final User user = oneLoginService.createOrGetUserFromInfo(userInfo);
-        addCustomJwtCookie(response, userInfo);
-        return new RedirectView(user.getLoginJourneyState()
-                .getLoginJourneyRedirect(user.getHighestRole().getName())
-                .getRedirectUrl(adminBaseUrl, applicantBaseUrl, techSupportAppBaseUrl, redirectUrlCookie));
+            final @RequestParam String code,
+            final @RequestParam String state
+    ) {
+        final JSONObject tokenResponse = oneLoginService.getOneLoginUserTokenResponse(code);
+        IdTokenDto decodedIdToken = oneLoginService.getDecodedIdToken(tokenResponse);
+        final String tokenNonce = decodedIdToken.getNonce();
+
+        final StateCookieDto stateCookieDto = oneLoginService.decodeStateCookie(stateCookie);
+        final String redirectUrl = stateCookieDto.getRedirectUrl();
+        final String cookieState = stateCookieDto.getState();
+        final String encodedStateJsonString = oneLoginService.buildEncodedStateJson(redirectUrl, cookieState);
+        final String hashedStateCookie = encryptionService.getSHA512SecurePassword(encodedStateJsonString);
+
+        final Nonce storedNonce = oneLoginService.readAndDeleteNonce(tokenNonce);
+        final Boolean nonceExpired = oneLoginService.isNonceExpired(storedNonce);
```

Comment:
personally I would suggest wrapping this in an if 
```Java
if(oneLoginService.isNonceExpired(storedNonce) 
{
// handle the expiration
}
else
{
//handle non-expired
}

```

Reply:
Currently we're treating expired and invalid the same way - will be building out capability to deactivate accounts at some point but until then we'll just redirect the user to log in again either way. I have refactored it to early-return an exception if expired/invalid which has cleaned up the /redirect-after-login function nicely